### ISSUE TITLE
Unsupervised Feature Learning and Deep Learning Tutorial

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10314]
Unsupervised Feature Learning and Deep Learning Tutorial<br>
<br>
https://ift.tt/OU6dufT<br>
<br>
<br>
<br>
<div><div>
<p>Consider a supervised learning problem where we have access to labeled training examples . Neural networks give a way of defining a complex, non-linear form of hypotheses , with parameters  that we can fit to our data.</p>
<p>To describe neural networks, we will begin by describing the simplest possible neural network, one which comprises a single “neuron.” We will use the following diagram to denote a single neuron:</p>
<center>
<img width="300" src="http://ufldl.stanford.edu/tutorial/images/SingleNeuron.png">
</center>
<p>This “neuron” is a computational unit that takes as input  (and a +1 intercept term), and outputs , where  is called the <strong>activation function</strong>. In these notes, we will choose  to be the sigmoid function:</p>

<p>Thus, our single neuron corresponds exactly to the input-output mapping defined by logistic regression.</p>
<p>Although these notes will use the sigmoid function, it is worth noting that another common choice for  is the hyperbolic tangent, or tanh, function:</p>

<p>Recent research has found a different activation function, the rectified linear function, often works better in practice for deep neural networks. This activation function is different from sigmoid and  because it is not bounded or continuously differentiable. The rectified linear activation function is given by,</p>

<p>Here are plots of the sigmoid,  and rectified linear functions:</p>
<center>
<img width="450" src="http://ufldl.stanford.edu/tutorial/images/Activation_functions.png">
</center>
<p>The  function is a rescaled version of the sigmoid, and its output range is  instead of . The rectified linear function is piece-wise linear and saturates at exactly 0 whenever the input  is less than 0.</p>
<p>Note that unlike some other venues (including the OpenClassroom videos, and parts of CS229), we are not using the convention here of . Instead, the intercept term is handled separately by the parameter .</p>
<p>Finally, one identity that’ll be useful later: If  is the sigmoid function, then its derivative is given by . (If  is the tanh function, then its derivative is given by .) You can derive this yourself using the definition of the sigmoid (or tanh) function. The rectified linear function has gradient 0 when  and 1 otherwise. The gradient is undefined at , though this doesn’t cause problems in practice because we average the gradient over many training examples during optimization.</p>
<h3>Neural Network model</h3>
<p>A neural network is put together by hooking together many of our simple “neurons,” so that the output of a neuron can be the input of another. For example, here is a small neural network:</p>
<center>
<img width="400" src="http://ufldl.stanford.edu/tutorial/images/Network331.png">
</center>
<p>In this figure, we have used circles to also denote the inputs to the network. The circles labeled “+1” are called <strong>bias units</strong>, and correspond to the intercept term. The leftmost layer of the network is called the <strong>input layer</strong>, and the rightmost layer the <strong>output layer</strong> (which, in this example, has only one node). The middle layer of nodes is called the <strong>hidden layer</strong>, because its values are not observed in the training set. We also say that our example neural network has 3 <strong>input units</strong> (not counting the bias unit), 3 <strong>hidden units</strong>, and 1 <strong>output unit</strong>.</p>
<p>We will let  denote the number of layers in our network; thus  in our example. We label layer  as , so layer  is the input layer, and layer  the output layer. Our neural network has parameters , where we write  to denote the parameter (or weight) associated with the connection between unit  in layer , and unit  in layer . (Note the order of the indices.) Also,  is the bias associated with unit  in layer . Thus, in our example, we have , and . Note that bias units don’t have inputs or connections going into them, since they always output the value +1. We also let  denote the number of nodes in layer  (not counting the bias unit).</p>
<p>We will write  to denote the <strong>activation</strong> (meaning output value) of unit  in layer . For , we also use  to denote the -th input. Given a fixed setting of the parameters , our neural network defines a hypothesis  that outputs a real number. Specifically, the computation that this neural network represents is given by:</p>

<p>In the sequel, we also let  denote the total weighted sum of inputs to unit  in layer , including the bias term (e.g., ), so that .</p>
<p>Note that this easily lends itself to a more compact notation. Specifically, if we extend the activation function  to apply to vectors in an element-wise fashion (i.e., ), then we can write the equations above more compactly as:</p>

<p>We call this step <strong>forward propagation.</strong> More generally, recalling that we also use  to also denote the values from the input layer, then given layer ’s activations , we can compute layer ’s activations  as:</p>

<p>By organizing our parameters in matrices and using matrix-vector operations, we can take advantage of fast linear algebra routines to quickly perform calculations in our network.</p>
<p>We have so far focused on one example neural network, but one can also build neural networks with other <strong>architectures</strong> (meaning patterns of connectivity between neurons), including ones with multiple hidden layers. The most common choice is a -layered network where layer  is the input layer, layer  is the output layer, and each layer  is densely connected to layer . In this setting, to compute the output of the network, we can successively compute all the activations in layer , then layer , and so on, up to layer , using the equations above that describe the forward propagation step. This is one example of a <strong>feedforward</strong> neural network, since the connectivity graph does not have any directed loops or cycles.</p>
<p>Neural networks can also have multiple output units. For example, here is a network with two hidden layers layers  and  and two output units in layer :</p>
<center>
<img width="500" src="http://ufldl.stanford.edu/tutorial/images/Network3322.png">
</center>
<p>To train this network, we would need training examples  where . This sort of network is useful if there’re multiple outputs that you’re interested in predicting. (For example, in a medical diagnosis application, the vector  might give the input features of a patient, and the different outputs ’s might indicate presence or absence of different diseases.)</p>
<h3>Backpropagation Algorithm</h3>
<p>Suppose we have a fixed training set  of  training examples. We can train our neural network using batch gradient descent. In detail, for a single training example , we define the cost function with respect to that single example to be:</p>

<p>This is a (one-half) squared-error cost function. Given a training set of  examples, we then define the overall cost function to be:</p>

<p>The first term in the definition of  is an average sum-of-squares error term. The second term is a regularization term (also called a <strong>weight decay</strong> term) that tends to decrease the magnitude of the weights, and helps prevent overfitting.</p>
<p>(Note: Usually weight decay is not applied to the bias terms , as reflected in our definition for . Applying weight decay to the bias units usually makes only a small difference to the final network, however. If you’ve taken CS229 (Machine Learning) at Stanford or watched the course’s videos on YouTube, you may also recognize this weight decay as essentially a variant of the Bayesian regularization method you saw there, where we placed a Gaussian prior on the parameters and did MAP (instead of maximum likelihood) estimation.)</p>
<p>The <strong>weight decay parameter</strong>  controls the relative importance of the two terms. Note also the slightly overloaded notation:  is the squared error cost with respect to a single example;  is the overall cost function, which includes the weight decay term.</p>
<p>This cost function above is often used both for classification and for regression problems. For classification, we let  or  represent the two class labels (recall that the sigmoid activation function outputs values in ; if we were using a tanh activation function, we would instead use -1 and +1 to denote the labels). For regression problems, we first scale our outputs to ensure that they lie in the  range (or if we were using a tanh activation function, then the  range).</p>
<p>Our goal is to minimize  as a function of  and . To train our neural network, we will initialize each parameter  and each  to a small random value near zero (say according to a  distribution for some small , say ), and then apply an optimization algorithm such as batch gradient descent. Since  is a non-convex function, gradient descent is susceptible to local optima; however, in practice gradient descent usually works fairly well. Finally, note that it is important to initialize the parameters randomly, rather than to all 0’s. If all the parameters start off at identical values, then all the hidden layer units will end up learning the same function of the input (more formally,  will be the same for all values of , so that  for any input ). The random initialization serves the purpose of <strong>symmetry breaking</strong>.</p>
<p>One iteration of gradient descent updates the parameters  as follows:</p>

<p>where  is the learning rate. The key step is computing the partial derivatives above. We will now describe the <strong>backpropagation</strong> algorithm, which gives an efficient way to compute these partial derivatives.</p>
<p>We will first describe how backpropagation can be used to compute  and , the partial derivatives of the cost function  defined with respect to a single example . Once we can compute these, we see that the derivative of the overall cost function  can be computed as:</p>

<p>The two lines above differ slightly because weight decay is applied to  but not .</p>
<p>The intuition behind the backpropagation algorithm is as follows. Given a training example , we will first run a “forward pass” to compute all the activations throughout the network, including the output value of the hypothesis . Then, for each node  in layer , we would like to compute an “error term”  that measures how much that node was “responsible” for any errors in our output. For an output node, we can directly measure the difference between the network’s activation and the true target value, and use that to define  (where layer  is the output layer). How about hidden units? For those, we will compute  based on a weighted average of the error terms of the nodes that uses  as an input. In detail, here is the backpropagation algorithm:</p>
<ol>
<li>
<p>Perform a feedforward pass, computing the activations for layers , , and so on up to the output layer .</p>
</li>
<li>
<p>For each output unit  in layer  (the output layer), set</p>

</li>
<li>
<p>For </p>
<p>    For each node  in layer , set</p>
<p>        </p>
</li>
<li>
<p>Compute the desired partial derivatives, which are given as:</p>
</li>
</ol>

<p>Finally, we can also re-write the algorithm using matrix-vectorial notation. We will use ”” to denote the element-wise product operator (denoted <code>.*</code> in Matlab or Octave, and also called the Hadamard product), so that if , then . Similar to how we extended the definition of  to apply element-wise to vectors, we also do the same for  (so that ).</p>
<p>The algorithm can then be written:</p>
<ol>
<li>
<p>Perform a feedforward pass, computing the activations for layers , , up to the output layer , using the equations defining the forward propagation steps</p>
</li>
<li>
<p>For the output layer (layer ), set</p>

</li>
<li>
<p>For , set</p>

</li>
<li>
<p>Compute the desired partial derivatives:</p>
</li>
</ol>

<p><strong>Implementation note:</strong> In steps 2 and 3 above, we need to compute  for each value of . Assuming  is the sigmoid activation function, we would already have  stored away from the forward pass through the network. Thus, using the expression that we worked out earlier for , we can compute this as .</p>
<p>Finally, we are ready to describe the full gradient descent algorithm. In the pseudo-code below,  is a matrix (of the same dimension as ), and  is a vector (of the same dimension as ). Note that in this notation, ”” is a matrix, and in particular it isn’t ” times .” We implement one iteration of batch gradient descent as follows:</p>
<ol>
<li>
<p>Set ,  (matrix/vector of zeros) for all .</p>
</li>
<li>
<p>For  to ,</p>
<ol>
<li>Use backpropagation to compute  and .</li>
<li>Set .</li>
<li>Set .</li>
</ol>
</li>
<li>
<p>Update the parameters:</p>
</li>
</ol>

<p>To train our neural network, we can now repeatedly take steps of gradient descent to reduce our cost function .</p>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via ufldl.stanford.edu https://ift.tt/Wqwz8Vg<br>
<br>
December 1, 2023 at 11:22AM
